File: src/main/java/com/github/encryptsl/magenta/MagentaLoader.java
package com.github.encryptsl.magenta;

import io.papermc.paper.plugin.loader.PluginClasspathBuilder;
import io.papermc.paper.plugin.loader.PluginLoader;
import io.papermc.paper.plugin.loader.library.impl.MavenLibraryResolver;
import org.eclipse.aether.artifact.DefaultArtifact;
import org.eclipse.aether.graph.Dependency;
import org.eclipse.aether.repository.RemoteRepository;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

@SuppressWarnings({"UnstableApiUsage", "unused"})
public class MagentaLoader implements PluginLoader {
    @Override
    public void classloader(@NotNull PluginClasspathBuilder pluginClasspath) {
        MavenLibraryResolver resolver = new MavenLibraryResolver();

        resolveLibraries().stream()
                .map(DefaultArtifact::new)
                .forEach(artifact -> resolver.addDependency(new Dependency(artifact, null)));
        resolver.addRepository(new RemoteRepository.Builder(
                "maven", "default", "https://repo.maven.apache.org/maven2/"
        ).build());

        resolver.addRepository(new RemoteRepository.Builder("paper", "default", "https://repo.papermc.io/repository/maven-public/").build());
        pluginClasspath.addLibrary(resolver);
    }

    private List<String> resolveLibraries() {
        try {
            return readLibraryListFromYaml();
        } catch (IOException e) {
            e.fillInStackTrace();
        }
        return new ArrayList<>();
    }

    private List<String> readLibraryListFromYaml() throws IOException {
        List<String> libraries = new ArrayList<>();
        InputStream inputStream = MagentaLoader.class.getClassLoader()
                .getResourceAsStream("paper-libraries.yml");

        if (inputStream != null) {
            try (inputStream; Scanner scanner = new Scanner(inputStream, StandardCharsets.UTF_8)) {
                while (scanner.hasNextLine()) {
                    libraries.add(scanner.nextLine());
                }
            }
        } else {
            System.err.println("paper-libraries.yml not found in the classpath.");
        }
        return libraries;
    }

}

File: src/main/kotlin/com/github/encryptsl/magenta/MagentaLoader.kt
package com.github.encryptsl.magenta

import io.papermc.paper.plugin.loader.PluginClasspathBuilder
import io.papermc.paper.plugin.loader.PluginLoader
import io.papermc.paper.plugin.loader.library.impl.MavenLibraryResolver
import org.eclipse.aether.artifact.DefaultArtifact
import org.eclipse.aether.graph.Dependency
import org.eclipse.aether.repository.RemoteRepository

class MagentaLoader : PluginLoader {

    override fun classloader(classBuilder: PluginClasspathBuilder) {
        val resolver = MavenLibraryResolver()
        resolver.addDependency(Dependency(DefaultArtifact("com.zaxxer:HikariCP:5.1.0"), null))
        resolver.addDependency(Dependency(DefaultArtifact("org.jetbrains.exposed:exposed-core:0.49.0"), null))
        resolver.addDependency(Dependency(DefaultArtifact("org.jetbrains.exposed:exposed-jdbc:0.49.0"), null))
        resolver.addDependency(Dependency(DefaultArtifact("org.jetbrains.exposed:exposed-kotlin-datetime:0.49.0"), null))
        resolver.addDependency(Dependency(DefaultArtifact("club.minnced:discord-webhooks:0.8.4"), null))
        resolver.addDependency(Dependency(DefaultArtifact("com.maxmind.geoip2:geoip2:4.2.0"), null))

        resolver.addRepository(RemoteRepository.Builder("paper", "default", "https://repo.papermc.io/repository/maven-public/").build())

        classBuilder.addLibrary(resolver)
    }

}